cram_3d_world/cram_bullet_reasoning/tests/robot-model-tests.lisp

```diff
@@ -0,0 +1,201 @@
+;;;
+;;; Copyright (c) 2016, Thomas Lipps <tlipps@uni-bremen.de>
+;;; All rights reserved.
+;;;
+;;; Redistribution and use in source and binary forms, with or without
+;;; modification, are permitted provided that the following conditions are met:
+;;;
+;;;     * Redistributions of source code must retain the above copyright
+;;;       notice, this list of conditions and the following disclaimer.
+;;;     * Redistributions in binary form must reproduce the above copyright
+;;;       notice, this list of conditions and the following disclaimer in the
+;;;       documentation and/or other materials provided with the distribution.
+;;;     * Neither the name of the Institute for Artificial Intelligence/
+;;;       Universitaet Bremen nor the names of its contributors may be used to
+;;;       endorse or promote products derived from this software without
+;;;       specific prior written permission.
+;;;
+;;; THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+;;; AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+;;; IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+;;; ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+;;; LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+;;; CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+;;; SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+;;; INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+;;; CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+;;; ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+;;; POSSIBILITY OF SUCH DAMAGE.
+(in-package :btr-tests)
+
+(defun setup-world ()
+  (setf rob-int:*robot-urdf*
+        (cl-urdf:parse-urdf (roslisp:get-param "robot_description")))
+  (prolog:prolog
+   `(and (btr:bullet-world ?world)
+         (rob-int:robot ?robot)
+         (assert (btr:object ?world :urdf ?robot ((0 0 0) (0 0 0 1))
+                             :urdf ,rob-int:*robot-urdf*))))
+  (btr:detach-all-objects (btr:get-robot-object)))
+
+(define-test attach-object-unknown-link
+  ;; Tries to attach an item to an unkown link of the robot. This should fail.
+  (setup-world)
+  (btr-utils:spawn-object 'o1 :mug :pose 
+                          '((-1 0.0 0.92)(0 0 0 1)))
+  (lisp-unit:assert-error 'simple-error
+                          (btr:attach-object (btr:get-robot-object) (btr:object btr:*current-bullet-world* 'o1)))
+  (lisp-unit:assert-error 'simple-error
+                          (btr:attach-object (btr:get-robot-object) (btr:object btr:*current-bullet-world* 'o1))
+                          :link "asdasdasd1")
+  (btr:remove-object btr:*current-bullet-world* 'o1))
+
+(define-test attach-object-same-object-to-two-links
```

Comment:
What happens, if an object is attached to the base_link and the gripper, and then the gripper is moved? Shouldn't an object be only attached to one other object/link?

Reply:
There seems to be a bug, if an item e.g. is attached twice to a robot. If the robot moves, the pose of the attached item will break. This will be fixed in another PR and therefore tested later. See Trello card https://trello.com/c/Iuu9zcs8/197-fix-broken-pose-of-attached-item-if-connected-twice-to-the-robot-eg-robot-holds-item-in-both-hands

Reply:
Should be fixed in https://github.com/cram2/cram/pull/112/commits/64e18c6fe86815d820fdd6941c257a039a43595e See above Trello Card.